uxarray/remap/nearest_neighbor.py

```diff
@@ -0,0 +1,200 @@
+                f"Invalid destination_data_mapping. Expected 'nodes' or 'face centers', "
+                f"but received: {destination_data_mapping}")
+
+        # specify whether to query on the corner nodes or face centers based on source grid
+        _source_tree = source_grid.get_ball_tree(tree_type=source_data_mapping)
+
+        # prepare coordinates for query
+        lonlat = np.vstack([lon, lat]).T
+
+        _, nearest_neighbor_indices = _source_tree.query(lonlat, k=1)
+
+        # data values from source data to destination data using nearest neighbor indices
+        if nearest_neighbor_indices.ndim > 1:
+            nearest_neighbor_indices = nearest_neighbor_indices.squeeze()
+
+        # support arbitrary dimension data using Ellipsis "..."
+        destination_data = source_data[..., nearest_neighbor_indices]
+
+        # case for 1D slice of data
+        if source_data.ndim == 1:
+            destination_data = destination_data.squeeze()
+
+        return destination_data
+
+    elif coord_type == "cartesian":
+        # TODO: once a cartesian balltree/kdtree is implemented, implement this
+        raise ValueError(
+            f"Nearest Neighbor Remapping using Cartesian coordinates is not yet supported"
+        )
+
+    else:
+        raise ValueError(
+            f"Invalid coord_type. Expected either 'lonlat' or 'Cartesian', but received {coord_type}"
+        )
+
+
+def _nearest_neighbor_uxda(source_uxda: UxDataArray,
+                           destination_obj: Union[Grid, UxDataArray, UxDataset],
+                           destination_data_mapping: str = "nodes",
+                           coord_type: str = "lonlat"):
+    """Nearest Neighbor Remapping implementation for ``UxDataArray``.
+
+    Parameters
+    ---------
+    source_uxda : UxDataArray
+        Source UxDataArray for remapping
+    destination_obj : Grid, UxDataArray, UxDataset
+        Destination for remapping
+    destination_data_mapping : str, default="nodes"
+        Location of where to map data, either "nodes" or "face centers"
+    coord_type : str, default="lonlat"
+        Indicates whether to remap using on latlon or Cartesian coordinates for nearest neighbor computations when
+        remapping.
+    """
+
+    # prepare dimensions
+    if destination_data_mapping == "nodes":
+        destination_dim = "n_node"
+    else:
+        destination_dim = "n_face"
+
+    destination_dims = list(source_uxda.dims)
+    destination_dims[-1] = destination_dim
+
+    if isinstance(destination_obj, Grid):
+        destination_grid = destination_obj
+    elif isinstance(
+            destination_obj,
+        (uxarray.core.dataarray.UxDataArray, uxarray.core.dataset.UxDataset)):
+        destination_grid = destination_obj.uxgrid
+    else:
+        raise ValueError("TODO: Invalid Input")
+
+    # perform remapping
+    destination_data = nearest_neighbor(source_uxda.uxgrid, destination_grid,
+                                        source_uxda.data,
```

Comment:
`.data` leads to an issue in `nearest_neighbor()` when `source_uxda` is a UxDataArray that wraps up a Dask array.

Reply:
Do you have any suggestions on what would be a better alternative?

Reply:
`.values` instead would guarantee the returned value to be a `np.ndarray` even if xarray was wrapping a Dask array. 

The big concern though in that case is that if the results of a really big data (which maybe wouldn't fit into memory) were being lazy-computed with a Dask array, we'd be forcing the data to be loaded into memory by calling `.values` explicitly. To avoid it, we'd need to write our methods in a way that they'd consume even internally Xarray.Dataarray instead of np.ndarray.

We need to make sure about this.

Reply:
Got it! Thank you for the clarification.

Reply:
Not only for this function but also for the whole code, as part of our performance and scalability work, I think we will need to check all our internal/user API functions from this point.

Reply:
Agreed, I've changed all instances within `UxDataArray` to use `.values` instead of `.data`, however I'll keep this in mind as we're working on our performance and optimization efforts.

Reply:
Sounds good!